### PR TITLE
Fixed #29049 -- Added slicing notation to F expressions.

### PR DESCRIPTION
**Pull request by priyanshsaxena**
F expressions on string-based fields and arrays can now use standard
python slicing-notation. Added a new class SlicableF that handles
identification of slicable fields. Step-argument is not supported.
Added unit-tests and documentation, with an entry in release-notes.

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -536,6 +539,60 @@ def _prepare(self, output_field=None):
...
+
+    def __repr__(self):
+        return "{}({})".format(self.__class__.__name__, self.expression.name)
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        text_fields = ['BinaryField', 'CharField', 'EmailField', 'FilePathField',
+                       'GenericIPAddressField', 'SlugField', 'TextField', 'URLField'
+                       ]
+        resolved = query.resolve_ref(self.expression.name, allow_joins, reuse, summarize)
+        if resolved.target.__class__.__name__ in text_fields:
```
Comment (atombrella) [inline django/db/models/expressions.py]: I think `get_internal_type` is better to use.

Comment (priyanshsaxena) [reply django/db/models/expressions.py]: Agreed. Didn't think of it. Fixed that.

File: django/db/models/expressions.py [hunk 2]
```diff
@@ -536,6 +539,61 @@ def _prepare(self, output_field=None):
         return self
 
 
+class SlicableF(F):
```
Comment (akulakov) [inline django/db/models/expressions.py]: spelling: `Sliceable`

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -536,6 +539,61 @@ def _prepare(self, output_field=None):
...
+        text_fields = ['BinaryField', 'CharField', 'EmailField', 'FilePathField',
+                       'GenericIPAddressField', 'SlugField', 'TextField', 'URLField'
+                       ]
+        resolved = query.resolve_ref(self.expression.name, allow_joins, reuse, summarize)
+        field_type = resolved.target.get_internal_type()
+        if field_type in text_fields:
+            from django.db.models.functions import Substr
+            return Substr(resolved, self.low, self.length)
+        elif field_type == "ArrayField":
+            from django.contrib.postgres.fields.array import SliceTransform
```
Comment (adamchainz) [inline django/db/models/expressions.py]: core django shouldn't be importing a contrib module, this makes it impossible for third party extensions impossible to hook into this functionality (some of the fields in my library Django-MySQL would want to use this functionality). Maybe there should be a function on the field class that can implement the kind of transform to apply?

File: django/db/models/expressions.py (hunk 2 above)
Comment (adamchainz) [reply django/db/models/expressions.py]: How about `SlicedF`? It's not 'sliceable' at this point, it has actually been sliced.

File: django/db/models/expressions.py (hunk 3 above)
Comment (adamchainz) [reply django/db/models/expressions.py]: Sorry I don't quite follow, a mixin for what class?

Comment (priyanshsaxena) [reply django/db/models/expressions.py]: Any class that derives from a new class called "Sliceable" can use the slicing operator. The fields can then be derived from this class as well to incorporate the required functionality.

It is also possible to add an attribute "is_sliceable"=False to the Field class. Any field that wants to allow for slicing can set this attribute to True. How would the slicing be implemented in that case is a question I don't have an answer to as yet.

**Comment by felixxm** [comment]
Updated in #11208.

**Pull request closed without merging**